### PR TITLE
Speed up n_body

### PR DESCRIPTION
**Pull request by llogiq**
This uses the algorithm from the packed_simd example without actually using SIMD – it relies fully on autovectorization. It's still a good deal faster than our current algorithm and I couldn't get it to work with stdsimd for now, so here we are.

before

```
real 0m5.905s
user 0m5.902s
sys  0m0.003s
```

after

```
real 0m3.745s
user 0m3.741s
sys  0m0.004s
```

@TeXitoi  Do you want to send it to Isaac or should I do it?

### CODE REVIEW

File: src/n_body.rs [hunk 1]
```diff
@@ -1,144 +1,224 @@
...
-    sun.vz = - pz / SOLAR_MASS;
-}
+pub fn advance(bodies: &mut [Body; N_BODIES], dt: f64) {
+    const N: usize = N_BODIES * (N_BODIES - 1) / 2;
 
-fn main() {
-    let n = std::env::args_os().nth(1)
-        .and_then(|s| s.into_string().ok())
-        .and_then(|n| n.parse().ok())
-        .unwrap_or(1000);
```
Comment (TeXitoi) [inline src/n_body.rs:135]: If no argument are given, it must use 1000. That's used in the tests.

File: src/n_body.rs [hunk 2]
```diff
@@ -1,144 +1,224 @@
 // The Computer Language Benchmarks Game
 // http://benchmarksgame.alioth.debian.org/
 //
-// contributed by the Rust Project Developers
-// contributed by Matt Brubeck
+// contributed by the packed_simd developers
```
Comment (TeXitoi) [inline src/n_body.rs:4]: Is their license compatible?

File: src/n_body.rs [hunk 3]
```diff
@@ -1,144 +1,224 @@
...
-fn energy(bodies: &[Planet;N_BODIES]) -> f64 {
-    bodies.iter().enumerate().fold(0.0, |e, (i, bi)| {
-        let ei = (bi.vx * bi.vx + bi.vy * bi.vy + bi.vz * bi.vz) * bi.mass / 2.0;
-        let ei2 = bodies[i + 1..].iter().fold(0.0, |ej, bj| {
+pub fn energy(bodies: &[Body; N_BODIES]) -> f64 {
+    let mut e = 0.;
+    for i in 0..N_BODIES {
+        let bi = &bodies[i];
+        e += bi.mass * (bi.v * bi.v).sum() * 0.5;
+        for bj in bodies.iter().take(N_BODIES).skip(i + 1) {
```
Comment (TeXitoi) [inline src/n_body.rs]: take?

File: src/n_body.rs [hunk 4]
```diff
@@ -1,144 +1,224 @@
...
+#[derive(Copy, Clone)]
+pub struct F64x2(f64, f64);
+
+#[derive(Copy, Clone)]
+pub struct F64x4(f64, f64, f64, f64);
+
+impl F64x2 {
+    pub fn splat(x: f64) -> F64x2 { F64x2(x, x) }
+    pub fn new(a: f64, b: f64) -> F64x2 { F64x2(a, b) }
+    pub fn sqrte(self) -> F64x2 { F64x2(self.0.sqrt(), self.1.sqrt()) }
```
Comment (TeXitoi) [inline src/n_body.rs]: Why a final e at sqrte?

File: src/n_body.rs [hunk 5]
```diff
@@ -1,144 +1,224 @@
...
-#[derive(Clone, Copy)]
-struct Planet {
-    x: f64, y: f64, z: f64,
-    vx: f64, vy: f64, vz: f64,
-    mass: f64,
+impl F64x4 {
+    pub fn splat(x: f64) -> F64x4 { F64x4(x, x, x, x) }
+    pub fn new(a: f64, b: f64, c: f64, d: f64) -> F64x4 { F64x4(a, b, c, d) }
+    pub fn sum(self) -> f64 {
+        (self.0 + self.1) + (self.2 + self.3)
```
Comment (TeXitoi) [inline src/n_body.rs:34]: Why these parenthesis? to minimize instructions?

**Review by TeXitoi — CHANGES_REQUESTED** [review]
Great!

**Comment by TeXitoi** [comment]
For the submission, as you wish. If you don't want to create a debian gitlab account, I already have one.

File: src/n_body.rs (hunk 4 above)
Comment (llogiq) [reply src/n_body.rs]: That's for 'estimate'. I can remove it, as it should no longer be one.

File: src/n_body.rs (hunk 5 above)
Comment (llogiq) [reply src/n_body.rs:34]: I found it beneficial to reduce data dependencies. Though the perf increase is admittedly small.

File: src/n_body.rs (hunk 1 above)
Comment (llogiq) [reply src/n_body.rs:135]: Ok, I'll change this back.

File: src/n_body.rs (hunk 2 above)
Comment (llogiq) [reply src/n_body.rs:4]: It's Apache 2/MIT, like Rust (as it's intended to move into `std` at some point). I believe this is compatible.

File: src/n_body.rs (hunk 3 above)
Comment (llogiq) [reply src/n_body.rs]: Ah, that's from the original – I'll reduce it to `&bodies[i +1..]`.

**Comment by llogiq** [comment]
Thanks @TeXitoi for the review. I think this should be good now. And since you have a gitlab account, I'll be glad to have you submit it.

**Review by TeXitoi — APPROVED** [review]

**Comment by TeXitoi** [comment]
https://salsa.debian.org/benchmarksgame-team/benchmarksgame/issues/42

**Comment by llogiq** [comment]
Thank you @TeXitoi !

**Comment by TeXitoi** [comment]
Isaac doesn't want macros.

**Comment by llogiq** [comment]
I'll expand the macros and re-submit.

**Pull request merged**